File: Practise/ArrayHashing/NextPermutation/nextperbf.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void nextPermuation(vector<int> &nums,vector<int> &temp,vector<vector<int>> &permutations,vector<bool> &visited)
{
    //Stopping condition
    if(temp.size() == nums.size())
    {
        permutations.push_back(temp);
        return;
    }

    // Traverse the decision space 
    for(int i = 0; i < nums.size(); i++)
    {
        // Traverse if not traversed 
        if(!visited[i])
        {
            // Make a choice
            temp.push_back(nums[i]);
            visited[i] = true;

            //Recurse
            nextPermuation(nums, temp, permutations, visited);

            //undo the choice
            temp.pop_back();
            visited[i] = false;
        }
    }

}

vector<int> genNextPermutation(vector<int> &nums)
{
    vector<int> temp;
    vector<bool> visited;
    for (int i = 0; i < nums.size();i++)
    {
        visited.push_back(false);
    }
    vector<vector<int>> permutations;

    vector<int> result;

    nextPermuation(nums, temp, permutations, visited);

    sort(permutations.begin(), permutations.end());

    for (int i = 0; i < permutations.size();i++)
    {
        if((nums == permutations[i]) && i < permutations.size() - 1)
        {
            result = permutations[i + 1];
            break;
        }
        else if((nums == permutations[i]) && i == permutations.size() - 1)
        {
            result = permutations[0];
        }
        return result;
    }
}

int main()
{
    // vector<int> some = {1, 2, 3, 4};
    // vector<int> some = {1,2,4,3};
    vector<int> some = {4,3,2,1};

    vector<int> result;
    result = genNextPermutation(some);

    for (int i = 0; i < result.size(); i++)
    {
        cout << result[i] << " ";
    }
    cout << endl;
}

File: Practise/ArrayHashing/NextPermutation/nextperbf.py
from typing import List

def nextPermutation(nums : List[int], temp : List[int], permutations : List[List[int]], visited : List[bool]) -> None :
    # Stopping condition : 
    if(len(nums) == len(temp)):
        permutations.append(temp[:])
        return
    
    
    for i in range(0,len(nums)):
        
        if (visited[i] == False):
            
            # Make a choice : 
            temp.append(nums[i])
            visited[i] = True
            
            # Recurse
            nextPermutation(nums,temp,permutations,visited)
            
            # Redo a choice
            temp.remove(nums[i])
            visited[i] = False
            
def genNextPermutation(nums : List[int]) -> List[int] :
    temp : List[int] = []
    permutations : List[List[int]] = []
    visited : List[bool]
    result : List[int] = []
    
    visited = [False for _ in range(len(nums))]
    
    nextPermutation(nums,temp,permutations,visited)
    
    permutations.sort()
    
    # print(permutations)
    for i in range(len(permutations)):
        if((nums == permutations[i]) and i < len(permutations) - 1):
            result = permutations[i+1][:]
            break
        elif ((nums == permutations[i]) and i == len(permutations) - 1):
            result = permutations[0][:]
    
    return result

def main():
    # some : List[int] = [1,2,3,4]
    # some : List[int] = [1,2,4,3]
    some : List[int] = [4,3,2,1]
    
    result : List[int] = genNextPermutation(some)
    
    print(result)
    
if __name__ == "__main__":
    main()
    
    
    